### ISSUE TITLE
Implement spawnProcess for IOCP reactor

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10774]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| justinj reported|
|-|-|
|Trac ID|trac#1008|
|Type|enhancement|
|Created|2005-04-29 23:26:22Z|

Attachments:

* [win32gui.pyd](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ff1/ff1eb8bd6cb17940ab78c0eeecf66268772f2061/7d59a19b441ff8feb09cda6e46a7b9603ad965d4.pyd) (94208 bytes) - added by justinj on 2005-08-03 03:16:05Z - 
* [final_patch2.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ff1/ff1eb8bd6cb17940ab78c0eeecf66268772f2061/1bcc941fc8e07c3e47de589ff34fd38b0139fc09.txt) (34274 bytes) - added by justinj on 2005-09-22 02:22:23Z - 
* [the_patch3.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ff1/ff1eb8bd6cb17940ab78c0eeecf66268772f2061/9eef0d6b0a3a1edb7773251827cf5e56495248f8.txt) (34301 bytes) - added by justinj on 2005-10-01 03:27:06Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1008 1008
type__enhancement enhancement
reporter__justinj justinj
priority__low low
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__win32 win32
time__1114817182000000 1114817182000000
changetime__1131588481000000 1131588481000000
version__ 
owner__justinj justinj
cc__Pahan cc__justinj cc__PenguinOfDoom cc__zooko
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Adding Pahan to Nosy List.  Is there any plan to implement this?  If not, are
there any recommendations for me if I were to start digging through the code to
see if I could tackle it?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Attached (iocp_spawnProcess.patch) is what I have so far, so I don't lose it.  
It had to be implemented with named pipes since anonymous pipes aren't support 
for async.  I'm not positive if my named pipe code is correct at this point.  
I'm also working out some details on how to call the appropriate methods on 
the Process transport in certain situations.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Updated patch attached (old one removed) as well as debugging info I'm 
currently using to track down a problem.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
The bug I'm currently getting (see FileMon.log) only happens when one or more 
of the initial three read operations returns CANCELLED status.  The closest 
thing I can find to a description of what that means is at 
http://www.cs.princeton.edu/~appel/smlnj/basis/posix-error.html which says 
cancelled indicates that the associated asynchronous operation was cancelled 
before completion.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Fixed!  New patch attached.  The only thing remaining is to implement getHost 
and getPeer.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
It was decided that getHost and getPeer aren't needed and don't make sense on 
a process transport.

After discussing on IRC, the task remaining is to wait for the process to end 
before getting the exit status instead of relying on the process having died 
after the 3 handles are closed.  Supposedly calls to GetExitCodeProcess will 
block at certain times if the process.  According to what exarkun thinks he 
remembers, there are 3 states: running, dying, dead.  If GetExitCodeProcess is 
called when the process is dying it can block.  It seems that our only option 
is to have a thread running WaitForMultipleObjects for each set of 64 
processes, which will notify the process transport when a process dies so we 
can safely get the exit status.  Ugly, but it seems there are no alternatives.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Attached (process.py) is my initial attempt at calling WaitForMultipleObjects 
on 64 process handles per thread.  It hasn't been tested and is dependent on 
win32gui.PeekMessage being added by Mark Hammond as he commented at 
http://mail.python.org/pipermail/python-win32/2005-July/003604.html.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Updated attached patch.  It passes all tests now.  Dependent on updated 
win32gui.pyd from Mark Hammond, until the next win32all release.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Updated patch.  Now unit test returns deferred, and wfmo threads properly 
close.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
See output.txt for the output I'm receiving when running...

python c:\python24\scripts\trial.py -r iocp 
twisted.test.test_process.ProcessTestCase.testManyProcesses
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
I am currently aware of two problems that occur when running the tests.

1) It seems that all64 processes in testManyProcesses get started and run and 
complete successfully, but _check never gets called and the test hangs and 
eventually returns with the following error:

[ERROR]: twisted.test.test_process.ProcessTestCase.testManyProcesses

  File "C:\src\Twisted\twisted\trial\util.py", 
line335, in wait
    r = _wait(d, timeout)
  File "C:\src\Twisted\twisted\trial\util.py", 
line282, in _wait
    raise defer.TimeoutError()
twisted.internet.defer.CancelledError:

2) Sometimes an error occurs because events happen in the wrong order and thus 
the stages list has the wrong contents.  I haven't seen this error in a 
while.  I'd like to resolve #1 first and see if this happens again.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
I think I may have fixed problem #1, and am running into #2 again.  I need to 
attach an updated patch, hopefully later tonight.  My code that closes the 
wfmo thread was in the wrong place.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Fixed out of order stages problem.  See updated patch.  I still occasionally 
get a lock up / timeout error.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Updated patch: cleaned up code and removed print statements.  Currently can't 
reproduce lock up problem, but I think it still exists.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
I reproduced the lockup problem and found out what is happening.  To notify 
the wfmo threads that to wait on new handles I call PostThreadMessage.  I 
added some code to track when a message was posted and when it was received 
and found that sometimes messages are sent but not received, resulting in wfmo 
eventually being called on an empty list which hangs.

Per PenguinOfDoom's recommendation I rewrote the code to use a Window to send 
messages using PostMessage instead of PostThreadMessage.  That code is now 
attached, but suffers from the same problem.  I'm not sure why this is 
happening.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Fixed.  Working patch attached as working_patch_using_win_msgs.txt
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Got code to hang again.  I fixed this though and after extensive testing have 
not been able to get it to hang.  Please see final_patch.txt and commit if 
you're okay with it.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
To be clear, I would like this code to be committed before the next release if 
at all possible.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Attached final_patch2.txt with some fixes to tests to remove log files.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
That last patch didn't include process.py.  Since process.py is new and I 
forgot to do an svn add of it before doing my diff.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Instead of being lazy I updated the patch to include process.py.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Yo!  Can you commit for me?  I keep trying to get a hold of you but we don't 
seem to sync up.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Updated patch with some additional test fixes.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Zooko, have you had time to look at the patch I committed to /branches/justinj?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)|@zooko commented|
|-|-|

```
#!html
<pre>
No, I haven't, but I still think that I will sometime soon after my current task
at work is complete.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Okay, please keep me updated.  Otherwise I'll just keep pestering you.  :)  
Thanks.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Any chance to look at this yet?
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Resolved in r14952.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Not all tests were run previously due to other bug which is now resolved.  
twisted.tests.test_iutils.testValue made a call to spawnProcess indirectly 
which resulted in an error dialog appearing on win32 build slave.  Reverted 
and will test on own branch.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="justinj's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|justinj commented|
|-|-|

```
#!html
<pre>
Fixed in r15084.
</pre>
```
